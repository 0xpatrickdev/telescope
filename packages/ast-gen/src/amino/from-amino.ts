import * as t from '@babel/types';
import { snake } from 'case';
import { Enum, Field, Interface, MessageSchema } from '../types';
import {
  BILLION,
  getFieldDimensionality,
  memberExpressionOrIdentifierSnake,
  shorthandProperty
} from '../utils';

export const fromAminoStringToLong = (prop: string, scope: string[]) => {
  return t.objectProperty(t.identifier(prop),
    t.callExpression(
      t.memberExpression(
        t.identifier('Long'),
        t.identifier('fromString')
      ),
      [
        memberExpressionOrIdentifierSnake(scope)
      ]
    ))
}

export const fromAminoFromArrayCall = (prop: string, scope: string[]) => {
  return t.objectProperty(t.identifier(prop),
    t.callExpression(
      t.memberExpression(
        t.identifier('Array'),
        t.identifier('from')
      ),
      [
        memberExpressionOrIdentifierSnake(scope)
      ]
    ))
}

export const fromAminoSnakeToCamel = (prop: string, scope: string[]) => {
  if (prop === snake(prop) && scope.length === 1) {
    return shorthandProperty(prop);
  }
  return t.objectProperty(t.identifier(prop), memberExpressionOrIdentifierSnake(scope))
}

export const fromAminoHeight = (prop: string, scope: string[]) => {
  return t.objectProperty(
    t.identifier(prop),
    t.conditionalExpression(
      memberExpressionOrIdentifierSnake(scope),
      t.objectExpression([
        t.objectProperty(t.identifier('revisionHeight'),
          t.callExpression(
            t.memberExpression(t.identifier('Long'), t.identifier('fromString')),
            [
              t.logicalExpression(
                '||',
                t.memberExpression(
                  memberExpressionOrIdentifierSnake(scope),
                  t.identifier('revision_height')
                ),
                t.stringLiteral('0')
              ),
              t.booleanLiteral(true)
            ])
        ),
        //
        t.objectProperty(t.identifier('revisionNumber'),
          t.callExpression(
            t.memberExpression(t.identifier('Long'), t.identifier('fromString')),
            [
              t.logicalExpression(
                '||',
                t.memberExpression(
                  memberExpressionOrIdentifierSnake(scope),
                  t.identifier('revision_number')
                ),
                t.stringLiteral('0')
              ),
              t.booleanLiteral(true)
            ])
        )
      ]),
      t.identifier('undefined')
    )
  )
};

export const fromAminoDuration = (prop: string, scope: string[]) => {
  const value = t.objectExpression(
    [
      t.objectProperty(t.identifier('seconds'), t.callExpression(
        t.memberExpression(t.identifier('Long'), t.identifier('fromNumber')), [
        t.callExpression(
          t.memberExpression(
            t.identifier('Math'),
            t.identifier('floor')
          ),
          [
            t.binaryExpression('/',
              t.callExpression(
                t.identifier('parseInt'),
                [
                  memberExpressionOrIdentifierSnake(scope)
                ]
              ),
              BILLION
            )
          ]
        )
      ]
      )),
      t.objectProperty(
        t.identifier('nanos'),
        t.binaryExpression('%',
          t.callExpression(
            t.identifier('parseInt'),
            [
              memberExpressionOrIdentifierSnake(scope)
            ]
          ),
          BILLION
        )
      )
    ]
  );
  return t.objectProperty(t.identifier(prop), value);
};

export const fromAminoEnum = (val: Enum, field: Field, scope: string[]) => {
  const toFunc = val.to.funcName;
  const value = t.callExpression(
    t.identifier(toFunc), [
    memberExpressionOrIdentifierSnake(scope)
  ]);
  return t.objectProperty(t.identifier(field.name), value);
};

export const fromAminoEnumArray = (val: Enum, field: Field, scope: string[]) => {
  const toFunc = val.to.funcName;
  const value = t.callExpression(
    t.memberExpression(
      memberExpressionOrIdentifierSnake(scope),
      t.identifier('map')
    ),
    [
      t.arrowFunctionExpression(
        [
          t.identifier('el')
        ],
        t.callExpression(
          t.identifier(toFunc),
          [
            t.identifier('el')
          ]
        )
      )
    ]
  );
  return t.objectProperty(t.identifier(field.name), value);
};

export const fromAminoInterface = (field: Field, ival: Interface, enums: Enum[], interfaces: Interface[], scope: string[], nested: number) => {
  const properties = ival.fields.map(field => {
    return fromAminoParseField(field, enums, interfaces, scope, nested + 1)
  });
  return t.objectProperty(t.identifier(field.name),
    t.objectExpression(
      properties
    )
  );
};

export const fromAminoInterfaceArray = (field: Field, ival: Interface, enums: Enum[], interfaces: Interface[], scope: string[], nested: number) => {
  const variable = 'el' + nested;
  const properties = ival.fields.map(field => {
    return fromAminoParseField(field, enums, interfaces, [variable], nested + 1)
  });

  const expr = t.callExpression(
    t.memberExpression(
      memberExpressionOrIdentifierSnake(scope),
      t.identifier('map')
    ),
    [
      t.arrowFunctionExpression(
        [
          t.identifier(variable)
        ],
        t.objectExpression(
          properties
        )
      )
    ]
  );

  return t.objectProperty(t.identifier(field.name),
    expr
  );
};

export const fromAminoParseField = (field: Field, enums: Enum[], interfaces: Interface[], scope = [], nested = 0) => {

  const newScope = [field.name, ...scope];
  const { typeName, dimensions, isArray } = getFieldDimensionality(field);

  // special cases
  switch (field.type) {
    case 'Duration':
      return fromAminoDuration(field.name, newScope);
    case 'Height':
      return fromAminoHeight(field.name, newScope);
    default:
  }

  // enums
  const val = enums.find(e => e.name === typeName);
  if (val) {
    if (!isArray) {
      return fromAminoEnum(val, field, newScope);
    } else if (dimensions === 0) {
      return fromAminoEnumArray(val, field, newScope);
    }
  }

  // definitions/interfaces
  const ival: Interface = interfaces.find(e => e.name === typeName);
  if (ival) {
    if (!isArray) {
      return fromAminoInterface(field, ival, enums, interfaces, newScope, nested);
    } else if (dimensions === 0) {
      return fromAminoInterfaceArray(field, ival, enums, interfaces, newScope, nested);
    }
  }

  switch (field.type) {
    case 'Long':
      return fromAminoStringToLong(field.name, newScope);
    case 'Coin[]':
      return fromAminoFromArrayCall(field.name, newScope);
    default:
      return fromAminoSnakeToCamel(field.name, newScope);
  }
};

export const fromAmino = (schema: MessageSchema, enums: Enum[], interfaces: Interface[]) => {

  const fromAminoParams = t.objectPattern(
    schema.fields.map((field) => t.objectProperty(t.identifier(snake(field.name)), t.identifier(snake(field.name)), false, true))
  );
  fromAminoParams.typeAnnotation = t.tsTypeAnnotation(t.tsIndexedAccessType(
    t.tsTypeReference(t.identifier('Amino' + schema.name)),
    t.tsLiteralType(t.stringLiteral('value'))
  ))

  const fromAmino = t.arrowFunctionExpression(
    [
      fromAminoParams
    ],
    t.blockStatement([
      t.returnStatement(
        t.objectExpression(
          schema.fields.map((field) => fromAminoParseField(field, enums, interfaces))
        )
      )
    ])
  );

  fromAmino.returnType = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(schema.name)));

  return fromAmino;
};
