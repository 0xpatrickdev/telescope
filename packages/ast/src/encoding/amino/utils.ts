import * as t from '@babel/types';
import { ProtoType } from '@cosmology/types';

export { getTypeUrlWithPkgAndName, getTypeUrl, getAminoTypeName } from '@cosmology/utils'

export const arrayTypeNDim = (body, n) => {
    if (!n || n <= 1) return t.tsArrayType(body);
    return t.tsArrayType(
        arrayTypeNDim(body, n - 1)
    );
};

export const protoFieldsToArray = (proto: ProtoType) => {
    return Object.keys(proto.fields).map(name => {
        return {
            name,
            ...proto.fields[name]
        };
    })
}