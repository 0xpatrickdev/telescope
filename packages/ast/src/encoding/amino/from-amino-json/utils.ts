import * as t from '@babel/types';
import { BILLION, memberExpressionOrIdentifier, memberExpressionOrIdentifierAminoCasing, shorthandProperty } from '../../../utils';
import { FromAminoParseField, fromAminoParseField } from './index'
import { protoFieldsToArray } from '../utils';
import { getOneOfs, getFieldOptionality } from '../../proto';

export const fromAmino = {
    defaultType(args: FromAminoParseField) {
        if (args.field.name === args.context.options.aminoCasingFn(args.field.name) && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.field.name), memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn))
    },

    string(args: FromAminoParseField) {
        if (args.field.name === args.context.options.aminoCasingFn(args.field.name) && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.field.name), memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn))
    },

    long(args: FromAminoParseField) {
        args.context.addUtil('Long');

        return t.objectProperty(t.identifier(args.field.name),
            t.callExpression(
                t.memberExpression(
                    t.identifier('Long'),
                    t.identifier('fromString')
                ),
                [
                    memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn)
                ]
            ));
    },

    duration(args: FromAminoParseField) {
        args.context.addUtil('Long');

        const value = t.objectExpression(
            [
                t.objectProperty(t.identifier('seconds'), t.callExpression(
                    t.memberExpression(t.identifier('Long'), t.identifier('fromNumber')), [
                    t.callExpression(
                        t.memberExpression(
                            t.identifier('Math'),
                            t.identifier('floor')
                        ),
                        [
                            t.binaryExpression('/',
                                t.callExpression(
                                    t.identifier('parseInt'),
                                    [
                                        memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn)
                                    ]
                                ),
                                BILLION
                            )
                        ]
                    )
                ]
                )),
                t.objectProperty(
                    t.identifier('nanos'),
                    t.binaryExpression('%',
                        t.callExpression(
                            t.identifier('parseInt'),
                            [
                                memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn)
                            ]
                        ),
                        BILLION
                    )
                )
            ]
        );
        return t.objectProperty(t.identifier(args.field.name), value);
    },

    height(args: FromAminoParseField) {
        args.context.addUtil('Long');

        return t.objectProperty(
            t.identifier(args.field.name),
            t.conditionalExpression(
                memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn),
                t.objectExpression([
                    t.objectProperty(t.identifier('revisionHeight'),
                        t.callExpression(
                            t.memberExpression(t.identifier('Long'), t.identifier('fromString')),
                            [
                                t.logicalExpression(
                                    '||',
                                    t.memberExpression(
                                        memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn),
                                        t.identifier(args.context.options.aminoCasingFn('revision_height'))
                                    ),
                                    t.stringLiteral('0')
                                ),
                                t.booleanLiteral(true)
                            ])
                    ),
                    //
                    t.objectProperty(t.identifier('revisionNumber'),
                        t.callExpression(
                            t.memberExpression(t.identifier('Long'), t.identifier('fromString')),
                            [
                                t.logicalExpression(
                                    '||',
                                    t.memberExpression(
                                        memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn),
                                        t.identifier(args.context.options.aminoCasingFn('revision_number'))
                                    ),
                                    t.stringLiteral('0')
                                ),
                                t.booleanLiteral(true)
                            ])
                    )
                ]),
                t.identifier('undefined')
            )
        )
    },

    enum({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField) {
        const enumFunction = context.lookupEnumFromJson(field, currentProtoPath);
        const value = t.callExpression(
            t.identifier(enumFunction), [
            memberExpressionOrIdentifierAminoCasing(scope, context.options.aminoCasingFn)
        ]);
        return t.objectProperty(t.identifier(field.name), value);
    },

    enumArray({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField) {
        const enumFunction = context.lookupEnumFromJson(field, currentProtoPath);
        const value = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifierAminoCasing(scope, context.options.aminoCasingFn),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier('el')
                    ],
                    t.callExpression(
                        t.identifier(enumFunction),
                        [
                            t.identifier('el')
                        ]
                    )
                )
            ]
        );
        return t.objectProperty(t.identifier(field.name), value);
    },

    type({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField) {
        const parentField = field;
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;
            return fromAminoParseField({
                context,
                field,
                currentProtoPath,
                scope: [...scope],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality?
            })
        });
        return t.objectProperty(t.identifier(field.name),
            t.objectExpression(
                properties
            )
        );
    },

    arrayFrom(args: FromAminoParseField) {
        return t.objectProperty(t.identifier(args.field.name),
            t.callExpression(
                t.memberExpression(
                    t.identifier('Array'),
                    t.identifier('from')
                ),
                [
                    memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn)
                ]
            ));
    },

    typeArray({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField) {
        const variable = 'el' + nested;
        const parentField = field;
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;

            return fromAminoParseField({
                context,
                field,
                currentProtoPath,
                scope: [variable],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality?
            })
        });

        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifierAminoCasing(scope, context.options.aminoCasingFn),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    t.objectExpression(
                        properties
                    )
                )
            ]
        );

        return t.objectProperty(t.identifier(field.name),
            expr
        );
    },


    scalarArray({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField, arrayTypeAstFunc: Function) {
        const variable = 'el' + nested;

        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifierAminoCasing(scope, context.options.aminoCasingFn),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    arrayTypeAstFunc(variable)
                )
            ]
        );

        return t.objectProperty(t.identifier(context.options.aminoCasingFn(field.name)),
            expr
        );
    },

    pubkey(args: FromAminoParseField) {
        args.context.addUtil('toBase64');
        args.context.addUtil('encodeBech32Pubkey');

        return t.objectProperty(
            t.identifier(args.field.name),
            t.callExpression(
                t.identifier('encodeBech32Pubkey'),
                [
                    t.objectExpression([
                        t.objectProperty(
                            t.identifier('type'),
                            t.stringLiteral('tendermint/PubKeySecp256k1')
                        ),
                        t.objectProperty(
                            t.identifier('value'),
                            t.callExpression(
                                t.identifier('toBase64'),
                                [
                                    t.memberExpression(
                                        t.identifier('pubkey'),
                                        t.identifier('value')
                                    )
                                ]
                            )
                        )
                    ]),
                    // TODO how to manage this?
                    // 1. options.prefix
                    // 2. look into prefix and how it's used across chains
                    // 3. maybe AminoConverter is a class and has this.prefix!
                    t.stringLiteral('cosmos')
                ]
            )
        )
    }
};


export const arrayTypes = {
    long(varname: string) {
        return t.callExpression(
            t.memberExpression(t.identifier('Long'), t.identifier('fromString')),
            [
                t.identifier(varname)
            ]
        )
    }
}
