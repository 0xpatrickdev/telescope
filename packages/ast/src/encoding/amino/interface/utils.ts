import * as t from '@babel/types';
import { memberExpressionOrIdentifierAminoCasing, shorthandProperty } from '../../../utils';
import { AminoOptions } from '../types';
import { arrayTypeNDim, BILLION } from '../utils';
import { getTypeFromContext, protoFieldsToArray } from '../utils';
import { getTSTypeFromProtoType, ProtoField } from '../../types';
import { RenderAminoField, renderAminoField } from '.';

export const aminoInterface = {
    defaultType(field: ProtoField, options: AminoOptions, ast: any) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(ast)
        );
    },
    long(field: ProtoField, options: AminoOptions) {
        // longs become strings...
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(t.tSStringKeyword())
        )
    },
    height(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier('AminoHeight'))
            )
        );
    },
    enum(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(t.tSNumberKeyword())
        );
    },
    enumArray(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(arrayTypeNDim(t.tSNumberKeyword(), 1))
        );
    },
    type({ context, field, options }: RenderAminoField) {
        const Type = getTypeFromContext(field, context);
        const properties = protoFieldsToArray(Type).map(field => {
            return renderAminoField({
                context,
                field,
                options
            })
        });

        // 
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                t.tsTypeLiteral(
                    properties
                )
            )
        );
    },
    typeArray({ context, field, options }: RenderAminoField) {
        const Type = getTypeFromContext(field, context);
        const properties = protoFieldsToArray(Type).map(field => {
            return renderAminoField({
                context,
                field,
                options
            })
        });

        // 
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                arrayTypeNDim(t.tsTypeLiteral(
                    properties
                ), 1)
            )
        );
    },

    array(field: ProtoField, options: AminoOptions) {
        // TODO write test case 

        // return t.tsPropertySignature(
        //     t.identifier(options.aminoCasingFn(field.name)),
        //     t.tsTypeAnnotation(
        //         arrayTypeNDim(t.tsTypeLiteral(
        //             properties
        //         ), 1)
        //     )
        // );
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                arrayTypeNDim(
                    getTSTypeFromProtoType(field.type),
                    1
                )
            )
        );
    }
}