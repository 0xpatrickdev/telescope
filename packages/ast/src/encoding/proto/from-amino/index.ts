import * as t from '@babel/types';
import { ProtoType, ProtoField } from '@osmonauts/types';
import { pascal } from 'case';
import { getFieldOptionality, getFieldOptionalityForDefaults, getOneOfs } from '..';
import { identifier, objectMethod } from '../../../utils';
import { ProtoParseContext } from '../../context';
import { fromAminoJSON, arrayTypes } from './utils';

const needsImplementation = (name: string, field: ProtoField) => {
    throw new Error(`need to implement fromAminoJSON (${field.type} rules[${field.rule}] name[${name}])`);
}
export interface FromAminoJSONMethod {
    context: ProtoParseContext;
    field: ProtoField;
    isOptional: boolean;
}

export const fromAminoJSONMethodFields = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const oneOfs = getOneOfs(proto);
    const fields = Object.keys(proto.fields ?? {}).map(fieldName => {
        const field = {
            name: fieldName,
            ...proto.fields[fieldName]
        };

        const isOneOf = oneOfs.includes(fieldName);
        const isOptional = getFieldOptionalityForDefaults(context, field, isOneOf);

        const args: FromAminoJSONMethod = {
            context,
            field,
            isOptional
        };

        if (field.rule === 'repeated') {
            switch (field.type) {
                case 'string':
                    return fromAminoJSON.array(args, arrayTypes.string());
                case 'bytes':
                    return fromAminoJSON.array(args, arrayTypes.bytes(args));
                case 'bool':
                    return fromAminoJSON.array(args, arrayTypes.bool());
                case 'float':
                    return fromAminoJSON.array(args, arrayTypes.float());
                case 'double':
                    return fromAminoJSON.array(args, arrayTypes.double());
                case 'int32':
                    return fromAminoJSON.array(args, arrayTypes.int32());
                case 'sint32':
                    return fromAminoJSON.array(args, arrayTypes.sint32());
                case 'uint32':
                    return fromAminoJSON.array(args, arrayTypes.uint32());
                case 'fixed32':
                    return fromAminoJSON.array(args, arrayTypes.fixed32());
                case 'sfixed32':
                    return fromAminoJSON.array(args, arrayTypes.sfixed32());
                case 'int64':
                    return fromAminoJSON.array(args, arrayTypes.int64());
                case 'sint64':
                    return fromAminoJSON.array(args, arrayTypes.sint64());
                case 'uint64':
                    return fromAminoJSON.array(args, arrayTypes.uint64());
                case 'fixed64':
                    return fromAminoJSON.array(args, arrayTypes.fixed64());
                case 'sfixed64':
                    return fromAminoJSON.array(args, arrayTypes.sfixed64());
                default:
                    switch (field.parsedType.type) {
                        case 'Enum':
                            return fromAminoJSON.array(args, arrayTypes.enum(args));
                        case 'Type':
                            return fromAminoJSON.array(args, arrayTypes.type(args));
                    }
                    return needsImplementation(fieldName, field);
            }
        }

        if (field.keyType) {
            switch (field.keyType) {
                case 'string':
                case 'int64':
                case 'sint64':
                case 'uint64':
                case 'fixed64':
                case 'sfixed64':
                case 'int32':
                case 'sint32':
                case 'uint32':
                case 'fixed32':
                case 'sfixed32':
                    return fromAminoJSON.keyHash(args);
                default:
                    return needsImplementation(fieldName, field);
            }
        }

        switch (field.type) {
            case 'string':
                return fromAminoJSON.string(args);
            case 'bytes':
                return fromAminoJSON.bytes(args);
            case 'bool':
                return fromAminoJSON.bool(args);
            case 'double':
                return fromAminoJSON.double(args);
            case 'float':
                return fromAminoJSON.float(args);
            case 'int32':
                return fromAminoJSON.int32(args);
            case 'sint32':
                return fromAminoJSON.sint32(args);
            case 'uint32':
                return fromAminoJSON.uint32(args);
            case 'fixed32':
                return fromAminoJSON.fixed32(args);
            case 'sfixed32':
                return fromAminoJSON.sfixed32(args);
            case 'int64':
                return fromAminoJSON.int64(args);
            case 'sint64':
                return fromAminoJSON.sint64(args);
            case 'uint64':
                return fromAminoJSON.uint64(args);
            case 'fixed64':
                return fromAminoJSON.fixed64(args);
            case 'sfixed64':
                return fromAminoJSON.sfixed64(args);
            case 'Duration':
            case 'google.protobuf.Duration':
                return fromAminoJSON.duration(args);
            case 'Timestamp':
            case 'google.protobuf.Timestamp':
                return fromAminoJSON.timestamp(args);
            default:
                switch (field.parsedType.type) {
                    case 'Enum':
                        return fromAminoJSON.enum(args);
                    case 'Type':
                        return fromAminoJSON.type(args);
                }
                return needsImplementation(fieldName, field);
        }
    });
    return fields;
};


export const fromAminoJSONMethod = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const fields = fromAminoJSONMethodFields(context, name, proto);
    let varName = 'object';
    if (!fields.length) {
        varName = '_';
    }

    const SDKTypeName =
        [name, 'SDKType']
            .filter(Boolean).join('');

    if (proto.type === 'Type' && proto.name === 'Duration') {
        // console.log(proto);
        // we could do something special here....
    }

    return objectMethod('method',
        t.identifier('fromAmino'),
        [
            identifier(varName,
                t.tsTypeAnnotation(
                    t.tsTypeReference(
                        t.identifier(SDKTypeName)
                    )
                ),
                false
            )

        ],
        t.blockStatement(
            [
                // t.variableDeclaration('const', [
                //     t.variableDeclarator(
                //         t.identifier('hoistedAsts'),
                //         t.nullLiteral()
                //     )
                // ]),

                t.returnStatement(
                    t.objectExpression(fields)
                )
            ]
        ),
        false,
        false,
        false,
        t.tsTypeAnnotation(
            t.tsTypeReference(
                t.identifier(name)
            )
        )
    )
};
