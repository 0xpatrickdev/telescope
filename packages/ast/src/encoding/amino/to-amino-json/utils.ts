import * as t from '@babel/types';
import { BILLION, memberExpressionOrIdentifier, shorthandProperty } from "../../../utils";
import { protoFieldsToArray } from '../utils';
import { ToAminoParseField, toAminoParseField } from './index'
import { getOneOfs, getFieldOptionality } from '../../proto';

export const toAmino = {
    defaultType(args: ToAminoParseField) {
        if (args.field.name === args.context.options.aminoCasingFn(args.field.name) && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), memberExpressionOrIdentifier(args.scope))
    },

    long(args: ToAminoParseField) {
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)),
            t.callExpression(
                t.memberExpression(memberExpressionOrIdentifier(args.scope), t.identifier('toString')),
                [])
        )
    },

    string(args: ToAminoParseField) {
        if (args.field.name === args.context.options.aminoCasingFn(args.field.name) && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), memberExpressionOrIdentifier(args.scope))
    },

    duration(args: ToAminoParseField) {
        const exp = t.binaryExpression(
            '*',
            memberExpressionOrIdentifier(args.scope),
            BILLION
        );
        exp.extra = { parenthesized: true };
        const value = t.callExpression(
            t.memberExpression(
                exp,
                t.identifier('toString')
            ),
            []
        )
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), value);
    },

    height(args: ToAminoParseField) {
        args.context.addUtil('omitDefault');

        const value = t.objectExpression(
            [
                t.objectProperty(
                    t.identifier(args.context.options.aminoCasingFn('revision_height')),
                    t.optionalCallExpression(
                        t.optionalMemberExpression(
                            t.callExpression(
                                t.identifier('omitDefault'),
                                [
                                    t.memberExpression(
                                        memberExpressionOrIdentifier(args.scope),
                                        t.identifier('revisionHeight')
                                    )
                                ]
                            ),
                            t.identifier('toString'),
                            false,
                            true
                        ),
                        [],
                        false
                    )
                ),
                //
                t.objectProperty(
                    t.identifier(args.context.options.aminoCasingFn('revision_number')),
                    t.optionalCallExpression(
                        t.optionalMemberExpression(
                            t.callExpression(
                                t.identifier('omitDefault'),
                                [
                                    t.memberExpression(
                                        memberExpressionOrIdentifier(args.scope),
                                        t.identifier('revisionNumber')
                                    )
                                ]
                            ),
                            t.identifier('toString'),
                            false,
                            true
                        ),
                        [],
                        false
                    )
                )
            ]
        );

        const cond = t.conditionalExpression(
            memberExpressionOrIdentifier(args.scope),
            value,
            t.objectExpression([])
        );

        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), cond);
    },

    coin(args: ToAminoParseField) {
        const value = t.objectExpression([
            t.objectProperty(t.identifier('denom'), t.memberExpression(
                memberExpressionOrIdentifier(args.scope),
                t.identifier('denom'),
            )),
            t.objectProperty(
                t.identifier('amount'),
                t.callExpression(
                    t.memberExpression(
                        t.callExpression(
                            t.memberExpression(
                                t.identifier('Long'),
                                t.identifier('fromNumber')
                            ),
                            [
                                t.memberExpression(
                                    memberExpressionOrIdentifier(args.scope),
                                    t.identifier('amount')
                                )
                            ]
                        ),
                        t.identifier('toString')
                    ),
                    []
                )
            )
        ]);
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), value);
    },

    type({ context, field, currentProtoPath, scope, nested, isOptional }: ToAminoParseField) {
        /// TODO (can this be refactored out? e.g. no recursive calls in this file?)
        /// BEGIN
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const parentField = field;
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;
            return toAminoParseField({
                context,
                field,
                currentProtoPath,
                scope: [...scope],
                nested,
                isOptional // TODO how to handle nested optionality
            })
        });
        /// END 
        return t.objectProperty(t.identifier(context.options.aminoCasingFn(field.name)),
            t.objectExpression(
                properties
            )
        );
    },


    typeArray({ context, field, currentProtoPath, scope, nested, isOptional }: ToAminoParseField) {
        const variable = 'el' + nested;

        if (field.parsedType.type !== 'Type') {
            throw new Error('Arrays only support types[Type] right now.');
        }

        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const parentField = field;
        const oneOfs = getOneOfs(Type);

        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;

            return toAminoParseField({
                context,
                field,
                currentProtoPath,
                scope: [variable],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality
            });
        });


        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifier(scope),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    t.objectExpression(
                        properties
                    )
                )
            ]
        );

        return t.objectProperty(t.identifier(context.options.aminoCasingFn(field.name)),
            expr
        );
    },

    pubkey(args: ToAminoParseField) {
        args.context.addUtil('fromBase64');
        args.context.addUtil('decodeBech32Pubkey');

        return t.objectProperty(
            t.identifier(args.field.name),
            t.objectExpression([
                t.objectProperty(
                    t.identifier('typeUrl'),
                    t.stringLiteral('/cosmos.crypto.secp256k1.PubKey')
                ),
                t.objectProperty(
                    t.identifier('value'),
                    t.callExpression(
                        t.identifier('fromBase64'),
                        [
                            t.memberExpression(
                                t.callExpression(
                                    t.identifier('decodeBech32Pubkey'),
                                    [
                                        t.identifier(args.field.name)
                                    ]
                                ),
                                t.identifier('value')
                            )

                        ]
                    )
                )
            ])
        )

    }


};
