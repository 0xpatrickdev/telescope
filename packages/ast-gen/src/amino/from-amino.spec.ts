import { memberExpressionOrIdentifier } from '../utils';
import { fromAminoSnakeToCamel } from './from-amino';
import generate from '@babel/generator';
import cases from 'jest-in-case';

const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}

cases('snake to camel', opts => {
    expectCode(memberExpressionOrIdentifier(opts.scope));
}, [
    { name: '1', prop: 'a', scope: ['a'] },
    { name: '2', prop: 'a', scope: ['a', 'b', 'c'] },
]);

cases('snake to camel', opts => {
    expectCode(fromAminoSnakeToCamel(opts.prop, opts.scope));
}, [
    { name: '1', prop: 'a', scope: ['a'] },
    { name: '2', prop: 'a', scope: ['a', 'b', 'c'] },
]);