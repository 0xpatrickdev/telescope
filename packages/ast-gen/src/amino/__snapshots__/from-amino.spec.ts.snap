// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snake to camel 1 1`] = `"a"`;

exports[`snake to camel 1 2`] = `"a"`;

exports[`snake to camel 2 1`] = `"c.b.a"`;

exports[`snake to camel 2 2`] = `"a: c.b.a"`;
